libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -0,0 +1,236 @@
+                QueryText = $"SELECT c.id, c.realId, c.document, c._etag FROM c WHERE c.id in ({parameterSequence})",
+                Parameters = new SqlParameterCollection(parameterValues)
+            };
+
+            var query = _client.CreateDocumentQuery<DocumentStoreItem>(collectionLink, querySpec).AsDocumentQuery();
+            while (query.HasMoreResults)
+            {
+                foreach (var doc in await query.ExecuteNextAsync<DocumentStoreItem>())
+                {
+                    var item = doc.Document.ToObject(typeof(object), _jsonSerializer);
+                    if (item is IStoreItem storeItem)
+                    {
+                        storeItem.eTag = doc.ETag;
+                    }
+
+                    // doc.Id cannot be used since it is escaped, read it from RealId property instead
+                    storeItems[doc.ReadlId] = item;
+                }
+            }
+
+            return storeItems;
+        }
+
+        /// <summary>
+        /// Persist the specified StoreItems into the CosmosDB Collection.
+        /// </summary>
+        /// <param name="changes">Items to persist.</param>
+        /// <returns></returns>
+        public async Task Write(StoreItems changes)
+        {
+            if (changes == null)
+            {
+                throw new ArgumentNullException(nameof(changes), "Please provide a StoreItems with changes to persist.");
+            }
+
+            var collectionLink = _collectionLink.Value;
+            foreach (var change in changes)
+            {
+                var json = JObject.FromObject(change.Value, _jsonSerializer);
+                json.Remove("eTag");
+                var documentChange = new DocumentStoreItem
+                {
+                    Id = SanitizeKey(change.Key),
+                    ReadlId = change.Key,
+                    Document = json
+                };
+
+                string eTag = (change.Value as IStoreItem)?.eTag;
+                if (eTag == null || eTag == "*")
+                {
+                    // if new item or * then insert or replace unconditionaly
+                    await _client.UpsertDocumentAsync(collectionLink, documentChange, disableAutomaticIdGeneration: true).ConfigureAwait(false);
+                }
+                else if (eTag.Length > 0)
+                {
+                    // if we have an etag, do opt. concurrency replace
+                    var uri = UriFactory.CreateDocumentUri(_databaseId, _collectionId, documentChange.Id);
+                    var ac = new AccessCondition { Condition = eTag, Type = AccessConditionType.IfMatch };
+                    await _client.ReplaceDocumentAsync(uri, documentChange, new RequestOptions { AccessCondition = ac }).ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new Exception("etag empty");
+                }
+            }
+        }
+
+        private string EnsureCollectionExist()
```

Comment:
Consider adding a comment describing that this is called as part of a Lazy<> resolution, and that thread safty concerns are thus minimized.